Support for compact blocks together with segwit

sipa opened the pull request:
See https://github.com/bitcoin/bips/pull/423 for BIP changes.

sipa committed:
Use cmpctblock type 2 for segwit-enabled transfer

Contains version negotiation logic by Matt Corallo and bugfixes by
Suhas Daftuar.

instagibbs commented:
nit: Might be worthwhile to print out the cmpctblk version being reconstructed in debug log

nit aside, utACK https://github.com/bitcoin/bitcoin/pull/8393/commits/1b417c20a6b0205a98268f95e6e19d27def076e9

btcdrak commented:
Tested ACK 1b417c2 (I tested this extensively this branch weeks before was opened).

NicolasDorier commented:
utACK 1b417c2 I write some tests with NBitcoin soon.

NicolasDorier commented:
`nLocalServices` is still never assigned `NODE_WITNESS` ?

NicolasDorier commented:
The days segwit will be activated, I think we'll get problems because nLocalService is assigned NODE_WITNESS only during init.

sdaftuar commented on src/main.cpp:
```diff
@@ -476,16 +479,16 @@ void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
-    if (nLocalServices & NODE_WITNESS) {
-        // Don't ever request compact blocks when segwit is enabled.
+    if (nLocalServices & ~pfrom->nServices & NODE_WITNESS) {
+        // Never ask from peers who can't provide witnesses.
```
This means that prior to segwit activation, blocks found by unupgraded miners will propagate without the benefit of compactblock announcements to upgraded nodes.

Perhaps it'd be better to instead use the BIP9 LOCKED_IN window to transition from receiving compactblock announcements from any peer regardless of nServices, towards NODE_WITNESS peers, and once ACTIVE state is reached, we enforce that all announcing peers are NODE_WITNESS?

To be concrete, here's one suggestion:
 1. before LOCKED_IN, use existing algorithm
 1. after LOCKED_IN, don't allow NODE_WITNESS peers to be evicted by non-WITNESS peers, and preferentially first evict non-WITNESS peers when trying to add a WITNESS peer
 1. after ACTIVATED, evict any non-WITNESS peers we have left making announcements.

I think this ought to provide a smooth transition, but I actually am not sure we even need that second step, we can probably be a little more disruptive around the LOCKED_IN time to ensure that we're ready at activation.

sipa commented:
Reminder: adapt to use the suggestion from BIP152 to just send both sendcmpct version 1 and version 2 (and let the 2 be ignored by pre-witness peers), rather than trying to guess what the peer will want based on service flags. Suggested by @TheBlueMatt

sipa commented:
Implemented the suggestion above.

sipa commented:
Fixed a bug found by @sdaftuar (see #8418).

jl2012 commented:
need backport to 0.13

sdaftuar commented on src/main.cpp:
```diff
@@ -5183,7 +5187,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             State(pfrom->GetId())->fProvidesHeaderAndIDs = !(nLocalServices & NODE_WITNESS) || (nCMPCTBLOCKVersion == 2);
             State(pfrom->GetId())->fPreferHeaderAndIDs = fAnnounceUsingCMPCTBLOCK;
-            State(pfrom->GetId())->fWantCmpctWitness = (nCMPCTBLOCKVersion == 2);
+            State(pfrom->GetId())->fSupportsCmpctWitness = (nCMPCTBLOCKVersion == 2);
```
This means that we'd overwrite `fSupportsCmpctWitness` if our peer sends a version 2 message followed by a version 1 message.  If we intend to require a certain ordering of sendcmpct announcements, I think that should be more clearly called out in the BIP (I'll leave a separate comment in bitcoin/bips#423 to indicate the language I suggest improving).  But my initial reaction is that we shouldn't impose this constraint.

Anyway I'm a bit confused on the interaction here between NODE_WITNESS and how we process sendcmpct messages.  Per BIP 152, if we don't support version 2 compact blocks, we must ignore version 2 sendcmpct messages.  But this code doesn't appear to be written this way (ie this processes version 2 sendcmpct messages, even if NODE_WITNESS is off).

In particular, it seems that this code could break if eg it was part of a release (like 0.13.0)  where NODE_WITNESS would not be set, but at some point in the future it had a peer that supported only version 2 compact blocks (as we'd process the version 2 message and think we could communicate with the peer, even though we can't).

sipa replied:
@sdaftuar I think they have to be sent in increasing order, so that the
receiver can use the simple "the last understood sendcmpct counts" logic.

You are right that we should ignore v2 sendcmpct messages when we don't
have NODE_WITNESS. That would make the behaviour of a non-defined segwit
node identical to a pre-segwit node. Agree?

sdaftuar replied:
@sipa  I guess that's fine.  I noted the language I thought could be clarified to make this more clear in the BIP.

I agree that ignoring v2 messages if we're not NODE_WITNESS should result in the identical behavior we're looking for.

sipa commented:
@sdaftuar Updated.

sipa commented:
Tested:
* mainnet (this PR) syncing from mainnet (this PR)
* testnet (this PR) syncing from testnet (this PR)
* testnet (this PR, with segwit activation disabled) syncing from testnet (this PR)
* testnet (post-CB, pre-segwit code) syncing from testnet (this PR)

They all seem to be using compact blocks for transfer.

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/8393/commits/00fabfeca09659f620b773ea5627c9a8ff4443cb

laanwj commented:
backport: this is for 0.13.1 I suppose?

sipa commented:
Yes, no need for this in 0.13.0.

sdaftuar commented:
@sipa Thoughts on this comment: https://github.com/bitcoin/bitcoin/pull/8393#discussion_r72060387 ?

sipa commented:
@sdaftuar Unsure... I think it's easier and more transparent to see failure/relay problems as early as possible, and rather at upgrade time than at lockedin time.

rebroad commented on src/main.cpp:
```diff
@@ -5087,14 +5091,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             bool fAnnounceUsingCMPCTBLOCK = false;
             uint64_t nCMPCTBLOCKVersion = 1;
             pfrom->PushMessage(NetMsgType::SENDCMPCT, fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion);
+            if (nLocalServices & NODE_WITNESS) {
```
Thanks. I find this much more readable and understandable than when developers create a function to do this. e.g. IsWitnessEnabled() !

sipa replied:
It is not the same. This condition is about whether a segwit softfork is defined, not whether it has activated.

rebroad commented on src/main.cpp:
```diff
@@ -5364,7 +5373,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             resp.txn[i] = block.vtx[req.indexes[i]];
         }
-        pfrom->PushMessageWithFlag(SERIALIZE_TRANSACTION_NO_WITNESS, NetMsgType::BLOCKTXN, resp);
+        pfrom->PushMessageWithFlag(State(pfrom->GetId())->fSupportsCmpctWitness ? 0 : SERIALIZE_TRANSACTION_NO_WITNESS, NetMsgType::BLOCKTXN, resp);
```
(pfrom->nServices & NODE_WITNESS) would, IMHO, be more readable, and clearer.

TheBlueMatt commented:
I proposed more specifics on version negotiation at https://github.com/bitcoin/bips/pull/423#issuecomment-244811877 which I believe might require slight implementation tweaks, though nothing significant.

sipa commented:
Added a commit that introduces the ping-pong based waiting described by the latest https://github.com/bitcoin/bips/pull/423.

cc @TheBlueMatt

TheBlueMatt commented:
I dont think you need the ping? As indicated in the new BIP text, you can know that the peer will respond to your request for a compact block with /at least/ a compact block of version X after you have seen at least that version in one of their announcements. For us, this is sufficient as we only care about them responding with version 2.

sdaftuar commented:
> Also, relay should work fine (and use compact blocks) as soon as a block from an unupgraded miner enters the network of upgraded peers, right? I assume that things like the relay network will even help propagation across that "boundary" to occur.

@sipa That sounds right to me.  I was thinking that the first miners to upgrade might be at a disadvantage (because they would stop receiving blocks from old miners as quickly as their peers), but I think you're right that the relay network should mitigate that.

sipa commented:
@TheBlueMatt I believe you're right, and the ping isn't necessary... but thinking about this makes my head hurt, so I feel better with an actual synchronization in place - even if we only need it later.

TheBlueMatt commented:
I removed the ping at https://github.com/TheBlueMatt/bitcoin/tree/segwitcb . While the proposed BIP text is pretty dense and confusing, the implementation is pretty straightforward: if they have sent us a sendcmpct version 2, we know they will encode things as version 2, so we gate actually requesting any compact blocks on having seen such a message. We also keep a boolean for which version they requested we send them.

TheBlueMatt commented:
Simplifed the impementation at the branch linked above and added a comment or two after @sdaftuar and @morcos looked over it a bit.

jonasschnelli commented:
Needs rebase.

sipa commented:
Reviewen and included @TheBlueMatt's suggested commits, and rebased. I have not tested this, and I likely won't have time for that this week.

TheBlueMatt commented:
Feel free to squash commits before release (you almost certainly want this since the different commits implement a different protocol)

sipa commented:
@TheBlueMatt Absolutely, I was planning to, especially as some commits revert changes from earlier ones.

sdaftuar reviewed (commented):
Code review ACK.  I'm working on updating the p2p test framework and p2p-compactblocks.py test to exercise the new behavior.

(Unfortunately I just discovered that one of the existing tests in p2p-compactblocks.py is pretty broken, working on a fix.)

sdaftuar committed:
[qa] Add support for compactblocks v2 to mininode

sdaftuar commented on src/main.cpp:
```diff
@@ -5194,12 +5211,22 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+            if (!State(pfrom->GetId())->fProvidesHeaderAndIDs) {
+                State(pfrom->GetId())->fProvidesHeaderAndIDs = true;
+                State(pfrom->GetId())->fWantsCmpctWitness = nCMPCTBLOCKVersion == 2;
+                State(pfrom->GetId())->fPreferHeaderAndIDs = fAnnounceUsingCMPCTBLOCK;
```
This prevents changing whether we announce to this peer.  It should be moved out, but I'm not sure yet what the condition should be for changing it.

sipa replied:
Cherry-picked the commit from https://github.com/TheBlueMatt/bitcoin/commits/segwitcb.

sdaftuar committed:
[qa] Update p2p-compactblocks.py for compactblocks v2

sdaftuar committed:
[qa] Fix bug in mininode witness deserialization

Also improve tx printing

instagibbs commented on src/main.cpp:
```diff
@@ -465,16 +478,16 @@ void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
 void MaybeSetPeerAsAnnouncingHeaderAndIDs(const CNodeState* nodestate, CNode* pfrom, CConnman& connman) {
-    if (pfrom->GetLocalServices() & NODE_WITNESS) {
-        // Don't ever request compact blocks when segwit is enabled.
+    if (!nodestate->fSupportsDesiredCmpctVersion) {
```
This condition is already checked before calling this function. Further it isn't clear how this stops our node from asking non-witness nodes to pre-emptively send cmpctblocks once segwit is activated.

TheBlueMatt replied:
A node is never prevented from chosing to send us a message, but later in this function is the only place we send a sendcmpct with announce set to true, so no matter how you call this it should be safe (note that we set fSupportsDesiredCmpctVersion based on NODE_WITNESS, not based on current activation status).

As for the duplicate check...whatever, up to the PR owner, I suppose :p.

instagibbs replied:
I failed to understand that pfrom->nLocalServices is talking about *your own* services. Nevermind.

sdaftuar commented:
What's the correct behavior if an old peer sends us a compact block message that is unrequested?  If I understand right, we're not currently checking to confirm that cmpctblock messages are only coming from peers with whom we've negotiated our preferred compact block version.

Accepting the header seems safe enough, and it's possible for the block to reconstruct correctly (eg if the block has no witness commitment), so there's perhaps no harm in trying to reconstruct, but it's probably not worth sending a GETBLOCKTXN message for missing transactions if the peer isn't using the preferred protocol version?

TheBlueMatt commented:
@sdaftuar I think we should emulate the block logic - if a peer sends us a block that is non-witness it should be the same case as if they send us a compact block message when we have negotiated to receive version 1, as that implies that block is a non-witness compact block.

instagibbs commented on src/main.cpp:
```diff
@@ -287,10 +287,21 @@ struct CNodeState {
+    //! Whether this peer wants witnesses in cmpctblocks/blocktxns
+    bool fWantsCmpctWitness;
+    /**
+     * For witness nodes: whether this peer sends witnesses in cmpctblocks/blocktxns
```
Maybe add "local" in both spots?

> For local witness nodes  

Semi ambiguous as-is.

TheBlueMatt replied:
local also seems unclear to me, how about "If we've announced NODE_WITNESS to this peer:" etc?

instagibbs replied:
ACK

instagibbs commented:
nit aside code review ACK https://github.com/bitcoin/bitcoin/pull/8393/commits/2048b69df36a5d1f49db2de6e099209631c1f490

sdaftuar commented:
ACK

If you'd like to address the minor issue I noted above about processing a compact block from a v1 peer post-segwit activation, I have a commit that will just do headers processing in such a case:
https://github.com/sdaftuar/bitcoin/commit/f6a86fb2112a4e7341bea6aaec0d4e55829e49c9

Also, I've reworked the p2p-compactblocks test for this PR.  The branch I've been working on is here:
https://github.com/sdaftuar/bitcoin/tree/test-8393-5.  If you want to include the tests that I wrote here from my branch, please feel free to take a look.  Note that one of the commits (https://github.com/sdaftuar/bitcoin/commit/e049ebf51fc3b8ab4f7a2f4ce19d8698ecf467bf) is a separate but necessary bugfix that I discovered in mininode.py, relating to witness deserialization.

Finally, the last two commits in that branch (https://github.com/sdaftuar/bitcoin/commit/d990722b56f02f3272c49eeef4c952992bd9efaa and https://github.com/sdaftuar/bitcoin/commit/2419b6b0256cff6ff9de1293e0cb87e2537e0f7f) may be of interest: I noticed it was a very small change to allow downloading compact blocks from old peers before segwit-activation (even if we don't ever ask such peers to announce compact blocks to us), so I coded that up and added a test for that case as well.

TheBlueMatt commented:
I'm in favor of adding something similar to the https://github.com/sdaftuar/bitcoin/commit/f6a86fb2112a4e7341bea6aaec0d4e55829e49c9, but prefer to enable on THRESHOLD_LOCKED_IN and not THRESHOLD_ACTIVE.

instagibbs commented:
@sdaftuar I'd ACK https://github.com/sdaftuar/bitcoin/commit/f6a86fb2112a4e7341bea6aaec0d4e55829e49c9 but I'm not sure the other changes are needed as those seem like unlikely corner cases to me that might happen just once and never again.

sipa commented:
I addressed @instagibbs's nits and cherry-picked @sdaftuar's proposed commits (I didn't review the tests yet). I'll think about the last two commits.

sipa commented:
I'd rather not pepper the block fetching code further with edge cases to deal with non-witness to witness propagation. The network should work fine without that, or we have other problems - and we'll know far in advance anyway.

Let me know if this should be squashed.

sipa commented:
Squashed version: https://github.com/sipa/bitcoin/commits/segwitcb_squashed

sdaftuar commented on src/main.cpp:
```diff
@@ -5690,6 +5690,12 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         CNodeState *nodestate = State(pfrom->GetId());
 
+        if (IsWitnessEnabled(pindex, chainparams.GetConsensus()) && !nodestate->fSupportsDesiredCmpctVersion) {
```
Oops.  This should actually be pindex->prev I think?  Sorry...

sdaftuar commented:
ACK after fixing the minor issue in my bugfix for the minor issue (sorry).  Feel free to squash.

sipa commented:
Nit addressed and squashed.

gmaxwell commented:
ACK

TheBlueMatt commented on src/main.cpp:
```diff
@@ -5757,7 +5791,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             // Might have collided, fall back to getdata now :(
             std::vector<CInv> invs;
-            invs.push_back(CInv(MSG_BLOCK, resp.blockhash));
+            invs.push_back(CInv(MSG_BLOCK | GetFetchFlags(pfrom, chainActive.Tip(), chainparams.GetConsensus()), resp.blockhash));
```
I believe this is wrong. If a witness peer announces, directly to us, via compact blocks, the first segwit block, and we are missing the second-to-first segwit block (or its currently in-flight) when we receive the response to this message, we will ban the peer sending the first segwit block as it will respond in non-witness form, because of this request.

sipa replied:
Does this still need changing, if we include #8871?

TheBlueMatt reviewed (commented):
There are a few places where we gate things on only fSupportsDesiredCmpctVersion, and not (LocalServices() & NODE_WITNESS) && fSupportsDesiredCmpctVersion (my fault, I wrote that), while this is still correct (since (LocalServices() & NODE_WITNESS) is a global constant after startup), it is somewhat strange and would break if we had to change something such that (LocalServices() & NODE_WITNESS) were no longer globally constant.
Don't really think its worth fixing, just thought it was worth a mention.

TheBlueMatt committed:
Make GetFetchFlags always request witness objects from witness peers

This fixes a bug where we might (in exceedingly rare circumstances)
accidentally ban a node for sending us the first (potentially few)
segwit blocks in non-segwit mode.

TheBlueMatt committed:
Fix overly-prescriptive p2p-segwit test for new fetch logic

TheBlueMatt commented:
With #8871 included first, utACK.

sipa commented:
Rebased on top of #8871.

sdaftuar commented:
ACK

petertodd commented:
utACK https://github.com/bitcoin/bitcoin/pull/8393/commits/f5b9b8f437c040205896ad0d7a6656efa08b5601

jonasschnelli commented:
utACK https://github.com/bitcoin/bitcoin/pull/8393/commits/f5b9b8f437c040205896ad0d7a6656efa08b5601
https://bitcoin.jonasschnelli.ch/pulls/8393/

btcdrak commented:
Post humus, re-tACK f5b9b8f437c040205896ad0d7a6656efa08b5601

rebroad commented on src/main.cpp:
```diff
@@ -5128,13 +5142,16 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+            if (pfrom->GetLocalServices() & NODE_WITNESS)
+                pfrom->PushMessage(NetMsgType::SENDCMPCT, fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion);
+            nCMPCTBLOCKVersion = 1;
             pfrom->PushMessage(NetMsgType::SENDCMPCT, fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion);
```
Seems a bit of a strange design that it needs to send two SENDCMPCT messages at once.... can it not send one OR the other instead?
